src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -204,440 +233,82 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
-                    return this.batchStream.ChangeSetBoundary == null
-                        ? ODataBatchReaderState.Completed
-                        : ODataBatchReaderState.ChangesetEnd;
-
-                case ODataBatchReaderState.ChangesetStart:
-                    // If we find an end boundary when in state 'ChangeSetStart' it means that
-                    // we have an empty changeset. The next state will be 'ChangeSetEnd'
-                    return ODataBatchReaderState.ChangesetEnd;
-
-                case ODataBatchReaderState.ChangesetEnd:
-                    // If we are at the end of a changeset and find an end boundary
-                    // we reached the end of the batch
-                    return ODataBatchReaderState.Completed;
-
-                case ODataBatchReaderState.Completed:
-                    // We should never get here when in Completed state.
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_Completed));
-
-                case ODataBatchReaderState.Exception:
-                    // We should never get here when in Exception state.
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_Exception));
-
-                default:
-                    // Invalid enum value
-                    throw new ODataException(Strings.General_InternalError(InternalErrorCodes.ODataBatchReader_GetEndBoundary_UnknownValue));
+                throw new ODataException(Strings.ODataBatchReader_MaxBatchSizeExceeded(this.inputContext.MessageReaderSettings.MessageQuotas.MaxPartsPerBatch));
             }
         }
 
         /// <summary>
-        /// Reads the next part from the batch message payload.
+        /// Increases the size of the current change set; throws if the allowed limit is exceeded.
         /// </summary>
-        /// <returns>true if more information was read; otherwise false.</returns>
-        private bool ReadSynchronously()
+        protected void IncreaseChangesetSize()
         {
-            return this.ReadImplementation();
+            this.currentChangeSetSize++;
+
+            if (this.currentChangeSetSize > this.inputContext.MessageReaderSettings.MessageQuotas.MaxOperationsPerChangeset)
+            {
+                throw new ODataException(Strings.ODataBatchReader_MaxChangeSetSizeExceeded(this.inputContext.MessageReaderSettings.MessageQuotas.MaxOperationsPerChangeset));
+            }
         }
 
-#if PORTABLELIB
         /// <summary>
-        /// Asynchronously reads the next part from the batch message payload.
+        /// Resets the size of the current change set to 0.
         /// </summary>
-        /// <returns>A task that when completed indicates whether more information was read.</returns>
-        private Task<bool> ReadAsynchronously()
+        protected void ResetChangesetSize()

```

Comment:
>ResetChangesetSize [](start = 23, length = 18)

Should be private and called from the base class once ReadImplementation is moved to the base class.

Reply:
IF we need to use the generic, basic framework of ReadImplementation similar to other (non-batch) JsonReader in base class, we need to take a hard look at the implementation of the JsonBatchReader to see how it can fit, some of the unique things we need to consider on the JSON side (see links below):
1. JsonReader has split implementations ReadImplementationForRequests and ReadImplementationForResponse to accommodate different shape of the response Json and request Json (note different attributes in request and response json, such as "status" only in response), and atomic groups.
2. atomic group processing is tied to the read implementation. Not sue at this point whether it can be implemented with the generic framework of ReadImplementation in base class.
3. Json processing requires caching, which is not the case for MIME batch reader.

Links to the JsonBatchReader on 6.x branch:
[https://github.com/biaol-odata/odata.net/blob/f88f8fd547a81cb51e975fe34311b3cbd7bdd4b9/src/Microsoft.OData.Core/JsonLight/ODataJsonLightBatchReader.cs#L189-L213](https://github.com/biaol-odata/odata.net/blob/f88f8fd547a81cb51e975fe34311b3cbd7bdd4b9/src/Microsoft.OData.Core/JsonLight/ODataJsonLightBatchReader.cs#L189-L213)
[https://github.com/biaol-odata/odata.net/blob/f88f8fd547a81cb51e975fe34311b3cbd7bdd4b9/src/Microsoft.OData.Core/JsonLight/ODataJsonLightBatchReader.cs#L555-L665](https://github.com/biaol-odata/odata.net/blob/f88f8fd547a81cb51e975fe34311b3cbd7bdd4b9/src/Microsoft.OData.Core/JsonLight/ODataJsonLightBatchReader.cs#L555-L665)


---
In reply to: [141155363](https://github.com/OData/odata.net/pull/960#discussion_r141155363) [](ancestors = 141155363)

Reply:
It's fine for the JsonBatchReader derived class to have separate logic within any given state for reading a request versus a response; it can use this.InputContext.ReadingResponse to determine if it is reading a response.  The ODataBatchReader should only be dealing with state transitions in calling the appropriate ReadAt[xyz]Implementation, and the existing contract is that those states must not differ depending on the format, or we have a bigger problem.